x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceMatcher.java

```diff
@@ -328,11 +332,13 @@ private void clearCircuitBreaker() {
     // expensive, so we just calculate the difference in bytes of the total memory that the matcher's
     // structure occupy for the in-flight tracking of sequences, as well as for the list of completed
     // sequences.
-    private void trackMemory(long prevRamBytesUsedInflight, long prevRamBytesUsedCompleted) {
-        long bytesDiff = ramBytesUsedInFlight() - prevRamBytesUsedInflight;
+    private void trackMemory() {
+        long bytesDiff = ramBytesUsedInFlight() - prevRamBytesUsedInFlight;
         addMemory(bytesDiff, CB_INFLIGHT_LABEL);
+        prevRamBytesUsedInFlight += bytesDiff;
```

Comment:
That's exactly the reason for this change: if we calculate the memory footprint change as the difference between the beginning and the end of `match()` execution, we risk to miss everything that happens outside of that.
Eg. 

- a match() is invoked, it calculates a delta of 10k: the real memory footprint is 10KB, the CB is updated with 10KB
- a trim() is invoked: the new real memory footprint is 0KB (the trim released all the memory), the CB is still 10KB (!!!) as `trim()` does not update the CB
- a match is invoked, it calculates another delta of 10k: the real memory footprint is 10KB again (the memory was cleared by the trim()), but the CB is updated to 20KB (!!!) because the new delta (10KB) is added to the CB value (10KB, not updated by trim)

Repeat this pattern for 100 times and at the end you'll have 10KB of real memory used, but the CB thinks you are using 1MB

The same could happen at the opposite, eg. something (outside of `match()` execution) allocates memory, but it does not update the CB (so the CB value is still zero!!!); after that, a `match()` execution REDUCES the memory footprint, so the delta is NEGATIVE. As soon as `match()` tries to update the CB with a negative value, you have the error we are chasing. 

Not sure how likely it is to happen with current implementation, but I see it as a weakness of current algorithm.

With a global variable (that is what this PR does) you solve all of this: you just calculate the total memory every time and you compare it with the total memory calculated at previous round. As long as Accountable objects cannot return negative values, you are always guaranteed that the CB cannot go below zero